forge-game/src/main/java/forge/game/trigger/Trigger.java

```diff
@@ -177,23 +170,43 @@ public final String replaceAbilityText(final String desc, SpellAbility sa) {
             sa = getOverridingAbility();
         }
         if (sa != null) {
-            String saDesc;
+            String saDesc = "";
+            boolean digMore = true;
             // if sa is a wrapper, get the Wrapped Ability
             if (sa.isWrapper()) {
                 final WrappedAbility wa = (WrappedAbility) sa;
                 sa = wa.getWrappedAbility();
 
-                // wrapped Charm spells are special,
-                // only get the selected abilities
+                // wrapped Charm spells are special, only get the selected abilities (if there are any yet)
                 if (ApiType.Charm.equals(sa.getApi())) {
                     saDesc = sa.getStackDescription();
-                } else {
-                    saDesc = sa.toString();
+                    digMore = false;
                 }
-            } else if (ApiType.Charm.equals(sa.getApi())) {
-                // use special formating, can be used in Card Description
-                saDesc = CharmEffect.makeFormatedDescription(sa);
-            } else {
+            }
+            if (digMore) { // if ABILITY is used, there is probably Charm somewhere
+                while (sa != null) {
+                    ApiType api = sa.getApi();
+                    if (ApiType.Charm.equals(api)) {
+                        saDesc = CharmEffect.makeFormatedDescription(sa, !forStack);
+                        break;
+                    }
+                    if (ApiType.ImmediateTrigger.equals(api) || ApiType.DelayedTrigger.equals(api)) {
+                        if (sa.hasAdditionalAbility("Execute")) {
+                            SpellAbility trigSA = sa.getAdditionalAbility("Execute");
```

Comment:
`getAdditionalAbility` already calls `hasAdditionalAbility` so you can save that if-block too
Should be good after